.github/workflows/CI.yml

```diff
@@ -85,13 +64,19 @@ jobs:
       - run: |
           python3 -m pip install --upgrade pipenv
           pipenv install --deploy --dev
-          pipenv run test
+          pipenv run test || TEST_FAILED=true
           pipenv run coverage xml
+          if [ "$TEST_FAILED" = true ]; then
+            echo "Tests failed, exiting step with error"
+            exit 1
+          fi
```

Comment:
IMO much simpler to create a different step to run the tests vs create the coverage report and upload it, then only one step needs the `if: always()` flag

Reply:
Updated